book/docs/lectures/Lecture4.md

```diff
@@ -0,0 +1,243 @@
+# Week04
+
+### Credits
+
+[Alberto Calzada - albertoCCz](https://github.com/albertoCCz)
+
+# The Contract Monad
+The porpouse of the Contract monad is to define off-chain code that runs in the wallet. It has four parameters:
+```haskell
+-- Contract w s e a
+```
+where:
+- a:  The overall result of the computation.
+- w:  Allows you to write log messages of type `w`. The equivalent of this would be the list of strings we had on the _Writer.hs_ example.
+- s:  Describes the blockchain capabilities aka what contract specific actions this contract can perform. For example: waiting for slots, submitting transactions, finding out your own public key or specific endpoints.
+- e:  Describes the type of error messages.
+
+### Simple example of Contract Monad
+In the next example we will just send a log message from the contract to the console, so we will be ignoring the `a` and `w` types.
+```haskell
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications  #-}
+-- ^ Add extensions on the top of the module file
+
+myContract :: Contract () BlockchainActions Text ()
+myContract = Contract.logInfo @String "Hello from the contract!"
+```
+As you can see, we have chosen `s` (which I believe stands for "schema") to be of type `BlockchainActions`. This data type contains the minimal set of actions for a contract: from the options given above, we will not be able to use specific endpoints. In particular, you can check what actions are available in the [Contract.hs](https://github.com/input-output-hk/plutus/blob/master/plutus-contract/src/Plutus/Contract.hs) module if you look for `BlockchainActions`.
+
+To use this contract you first need to define the trace, which substitutes what we previouly did on the Plutus Playground by providing with a list of actions the wallet(s) associated to the contract is going to perfom. The trace function we are going to use is the next one:
+```haskell
+myTrace :: EmulatorTrace ()
+myTrace = void $ activateContractWallet (Wallet 1) myContract
+```
+This function just activates a wallet or set of wallets (Wallet 1 in this case) by associating it with a contract (myContract in this case) and normally the result is saved in a handler that we can use later. Normally, the code would look like this:
+```haskell
+myTrace :: EmulatorTrace ()
+myTrace = h <- activateContractWallet (Wallet 1) myContract
+```
+where `h` is the handler. Now, as we are just interested in showing the log message and we will not use the handler, we use the `void` keyword so the compiler does not complaint. Finally, after defining the trace, we can define one or more test(s) to study whether or not the contract works as expected. This time we only define one of these tests because, again, we are just interested in the log message. The test is:
```

Comment:
```suggestion
where `h` is the handler. Now, since we are only interested in showing the log message and not using the handler, we will use the `void` keyword so the compiler does not complain. Finally, after defining the trace, we can define one or more test(s) to study whether or not the contract works as expected. This time we only define one of these tests because, again, we are just interested in the log message. The test is:
```